src/main/java/net/greghaines/jesque/utils/JedisUtils.java

```diff
@@ -175,4 +188,30 @@ public static boolean canUseAsRecurringQueue(final Jedis jedis, final String que
     private JedisUtils() {
         // Utility class
     }
+
+    /**
+     * Determines if the queue identified by the given key can be used as a priority queue.
+     *
+     * @param jedis connection to Redis
+     * @param key   the key that identifies a queue  @return true if the key already is a delayed queue or is not currently used, false otherwise
+     */
+    public static boolean canUseAsPriorityQueue(final Jedis jedis, final String key) {
+        return useAsQueueOfType(jedis, key, PRIORITY_QUEUE);
+    }
+
+    private static boolean useAsQueueOfType(final Jedis jedis, final String key, final String queueType) {
+        final String existingType = jedis.hget(QUEUE_TYPES, key);
+        if (existingType == null) {
+            jedis.hset(QUEUE_TYPES, key, queueType);
```

Comment:
Thank you for the quick reply.
You are right about the race condition but I believe it also exists without
the modification.
client A calls `canUseAsDelayedQueue`, client B calls `enqueue`, client A
calls `jedis.zadd` on list etc'.

I pushed a fix that maintains the current semantics of reassigning the
queue type as long as it is empty.

But isn't interoperability with Ruby and node.js a problem?


On Thu, Jan 7, 2016 at 2:20 AM, Greg Haines <notifications@github.com>
wrote:

> In src/main/java/net/greghaines/jesque/utils/JedisUtils.java
> <https://github.com/gresrun/jesque/pull/95#discussion_r49028507>:
>
> > @@ -175,4 +188,30 @@ public static boolean canUseAsRecurringQueue(final Jedis jedis, final String que
> >      private JedisUtils() {
> >          // Utility class
> >      }
> > +
> > +    /**
> > +     * Determines if the queue identified by the given key can be used as a priority queue.
> > +     *
> > +     * @param jedis connection to Redis
> > +     * @param key   the key that identifies a queue  @return true if the key already is a delayed queue or is not currently used, false otherwise
> > +     */
> > +    public static boolean canUseAsPriorityQueue(final Jedis jedis, final String key) {
> > +        return useAsQueueOfType(jedis, key, PRIORITY_QUEUE);
> > +    }
> > +
> > +    private static boolean useAsQueueOfType(final Jedis jedis, final String key, final String queueType) {
> > +        final String existingType = jedis.hget(QUEUE_TYPES, key);
> > +        if (existingType == null) {
> > +            jedis.hset(QUEUE_TYPES, key, queueType);
>
> @itaim <https://github.com/itaim> I believe there is a race condition
> here if two clients are attempting to create a queue of the same name but
> different type at the same time. Also, it looks like we never remove
> entries from this hash, which conflicts with the current semantics of
> reassigning the queue type as long as it is empty.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gresrun/jesque/pull/95/files#r49028507>.
>



-- 
*Itai Marks*
Architect // *Kenshoo*
*Office* +972 732862390 // *Mobile* +972 526261166

-- 
This e-mail, as well as any attached document, may contain material which 
is confidential and privileged and may include trademark, copyright and 
other intellectual property rights that are proprietary to Kenshoo Ltd, 
 its subsidiaries or affiliates ("Kenshoo"). This e-mail and its 
attachments may be read, copied and used only by the addressee for the 
purpose(s) for which it was disclosed herein. If you have received it in 
error, please destroy the message and any attachment, and contact us 
immediately. If you are not the intended recipient, be aware that any 
review, reliance, disclosure, copying, distribution or use of the contents 
of this message without Kenshoo's express permission is strictly prohibited.